stability_image/image.go

```diff
@@ -0,0 +1,392 @@
+package stability_image
+
+import (
+	"bytes"
+	"github.com/disintegration/imaging"
+	"github.com/esimov/stackblur-go"
+	"github.com/foobaz/lossypng/lossypng"
+	"github.com/lucasb-eyer/go-colorful"
+	"github.com/mazznoer/colorgrad"
+	"golang.org/x/image/webp"
+	"image"
+	"image/color"
+	"image/gif"
+	"image/jpeg"
+	"image/png"
+	"log"
+	"math/rand"
+	"os"
+	"strconv"
+)
+
+var (
+	MaxDimension     uint64 = 1536
+	MinDimension     uint64 = 256
+	MaxPixels        uint64 = 960 * 1024
+	MinPixels        uint64 = 768 * 768
+	DefaultDimension uint64 = 768
+	DimensionStep    uint64 = 64
+)
+
+func EncodePng(
+	i image.Image,
+	level png.CompressionLevel,
+) (
+	encoded *[]byte,
+	err error,
+) {
+	encoded = &[]byte{}
+	buf := new(bytes.Buffer)
+
+	encoder := png.Encoder{CompressionLevel: level}
+	writeErr := encoder.Encode(buf, i)
+	if writeErr != nil {
+		return nil, err
+	}
+	*encoded = buf.Bytes()
+	return encoded, nil
+}
+
+func DecodeImage(raw *[]byte) (
+	i image.Image,
+	format string,
+	dim *image.Point,
+	err error) {
+	i, format, err = image.Decode(bytes.NewReader(*raw))
+	if err != nil {
+		return nil, "", dim, err
+	}
+	dimVal := i.Bounds().Size()
+	return i, format, &dimVal, err
+}
+
+func Fill(img *image.RGBA, col *color.RGBA) {
+	bounds := img.Bounds()
+	for x := bounds.Min.X; x < bounds.Max.X; x++ {
+		for y := bounds.Min.Y; y < bounds.Max.Y; y++ {
+			img.Set(x, y, col)
+		}
+	}
+}
+
+// CreateNoiseImage creates a noise image with the given dimensions.
+func CreateNoiseImage(width int, height int) (i *image.RGBA, err error) {
+	i = image.NewRGBA(image.Rect(0, 0, width, height))
+	for x := 0; x < i.Rect.Max.X; x++ {
+		for y := 0; y < i.Rect.Max.Y; y++ {
+			i.Set(x, y, color.RGBA{
+				R: uint8(rand.Intn(255)),
+				G: uint8(rand.Intn(255)),
+				B: uint8(rand.Intn(255)),
+				A: 255,
+			})
+		}
+	}
+	return i, nil
+}
+
+// ShufflePixelsImage shuffles the pixels of an image within a given bounds.
+func ShufflePixelsImage(i *image.NRGBA, bounds image.Rectangle) {
```

Comment:
It looks like this assumes bounds.Min.X == 0 and bounds.Min.Y == 0 otherwise it will be shuffling with pixels outside the rectangle. Is this the intent or is it supposed to be restricted to writing within the given bounds?